### PR TITLE
Move some QNN EP provider options to session options

### PR DESCRIPTION
**Pull request by HectorSVC**
Move QNN EP provider options to session options

### Description
Need to use session option to support multi-partition for context cache feature. To smooth the transaction, move the provider options to session options first.

This is the first step for PR:
PR https://github.com/microsoft/onnxruntime/pull/18865

### CODE REVIEW

File: onnxruntime/test/perftest/ort_test_session.cc [hunk 1]
```diff
@@ -272,7 +272,7 @@ OnnxRuntimeTestSession::OnnxRuntimeTestSession(Ort::Env& env, std::random_device
         } else {
           ORT_THROW("[ERROR] [OpenVINO] The value for the key 'enable_opencl_throttling' should be a boolean i.e. true or false. Default value is false.\n");
         }
-      } else if (key == "disable_dynamic_shapes") {
+      } else if (key == "enable_dynamic_shapes") {
```
Comment (jywu-msft) [inline onnxruntime/test/perftest/ort_test_session.cc:275]: i think you picked up some openvino changes that you didn't mean to?

File: onnxruntime/test/perftest/ort_test_session.cc [hunk 2]
```diff
@@ -298,7 +298,7 @@ OnnxRuntimeTestSession::OnnxRuntimeTestSession(Ort::Env& env, std::random_device
           ov_options[key] = value;
         }
       } else {
-        ORT_THROW("[ERROR] [OpenVINO] wrong key type entered. Choose from the following runtime key options that are available for OpenVINO. ['device_type', 'device_id', 'enable_npu_fast_compile', 'num_of_threads', 'cache_dir', 'num_streams', 'enable_opencl_throttling', 'disable_dynamic_shapes'] \n");
+        ORT_THROW("[ERROR] [OpenVINO] wrong key type entered. Choose from the following runtime key options that are available for OpenVINO. ['device_type', 'device_id', 'enable_npu_fast_compile', 'num_of_threads', 'cache_dir', 'num_streams', 'enable_opencl_throttling|true'] \n");
```
Comment (jywu-msft) [inline onnxruntime/test/perftest/ort_test_session.cc:301]: same here. this is not related to your change right

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 3]
```diff
@@ -235,3 +235,18 @@
 // Use this config to control the minimum size of the initializer when externalizing it during serialization
 static const char* const kOrtSessionOptionsOptimizedModelExternalInitializersMinSizeInBytes =
     "session.optimized_model_external_initializers_min_size_in_bytes";
+
+// Enable EP context feature to dump the partitioned graph which include the EP context into Onnx file.
+// The dumped Onnx model with EP context can be used for future inference to avoid the EP graph partitioning/compile overhead.
+// "0": disable. (default)
+// "1": enable.
+static const char* const kOrtSessionOptionEpContextEnable = "ep.context_enable";
```
Comment (HectorSVC) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:243]: need to add CPI for these new options

Comment (HectorSVC) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:243]: realized that AddSessionConfigEntry can do the job

---
In reply to: [1431842901](https://github.com/microsoft/onnxruntime/pull/18877#discussion_r1431842901) [](ancestors = 1431842901)

**Comment by jywu-msft** [comment]
I assume you'll update the inference-examples https://github.com/microsoft/onnxruntime-inference-examples/blob/25df47480be349482ea2c6efe12017236e0fe172/c_cxx/QNN_EP/mobilenetv2_classification/main.cpp#L89
and docs https://onnxruntime.ai/docs/execution-providers/QNN-ExecutionProvider.html ?

**Review by jywu-msft — APPROVED** [review]

**Review by adrianlizarraga — DISMISSED** [review]

**Comment by HectorSVC** [comment]
inference example PR: https://github.com/microsoft/onnxruntime-inference-examples/pull/354

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h (hunk 3 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:243]: This seems specific to QNN so a prefix of 'qnn.' instead of 'session.' will be much clearer

Comment (HectorSVC) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:243]: It can be re-used by other EPs like TensorRT EP.

---
In reply to: [1432239944](https://github.com/microsoft/onnxruntime/pull/18877#discussion_r1432239944) [](ancestors = 1432239944)

Comment (jywu-msft) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:243]: > It can be re-used by other EPs like TensorRT EP.
> 
> In reply to: [1432239944](https://github.com/microsoft/onnxruntime/pull/18877#discussion_r1432239944) [](ancestors = 1432239944)

it will be used by TensorRT EP. Currently it also uses EPContext, but to support more than one partition, we will need this as well.

**Review by jywu-msft — APPROVED** [review]

File: onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc [hunk 4]
```diff
@@ -160,7 +160,7 @@ bool IsContextCacheFileExists(const std::string& customer_context_cache_path,
   if (!customer_context_cache_path.empty()) {
     context_cache_path = ToPathString(customer_context_cache_path);
   } else if (!model_pathstring.empty()) {
-    context_cache_path = model_pathstring + ToPathString("_qnn_ctx.onnx");
+    context_cache_path = model_pathstring + ToPathString("_ctx.onnx");
```
Comment (jywu-msft) [inline onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc:163]: since this particular implementation is for qnn, i'm wondering if we should keep qnn in the name?

Comment (HectorSVC) [reply onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc:163]: Eventually, we will support multi-partition and move this model saving to inference session level. We do need to remove EP specific label.
I remove this qnn label and updated the inference code accordingly so that we don't need to change the inference sample code again once the multi-partition support get in.

---
In reply to: [1432352771](https://github.com/microsoft/onnxruntime/pull/18877#discussion_r1432352771) [](ancestors = 1432352771)

Comment (HectorSVC) [reply onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc:163]: Also, this is the default name in case users do not set it. User should set their own name normally.

---
In reply to: [1432369068](https://github.com/microsoft/onnxruntime/pull/18877#discussion_r1432369068) [](ancestors = 1432369068,1432352771)
